trie/zk/merkle_tree.go

```diff
@@ -0,0 +1,481 @@
+) (*ParentNode, error) {
+	if lvl > t.maxLevels-2 {
+		return nil, trie.ErrReachedMaxLevel
+	}
+	if pathNewLeaf.Get(lvl) == pathOldLeaf.Get(lvl) { // We need to go deeper!
+		nextNode, err := t.pushLeaf(newLeaf, oldLeaf, lvl+1, pathNewLeaf, pathOldLeaf)
+		if err != nil {
+			return nil, err
+		}
+		newParentNode := NewParentNode()
+		newParentNode.SetChild(pathNewLeaf.Get(lvl), nextNode)
+		newParentNode.SetChild(pathNewLeaf.GetOther(lvl), EmptyNodeValue)
+		return newParentNode, t.checkNodeHashConflictDuringUpdate(newParentNode)
+	}
+	if err := t.checkNodeHashConflictDuringUpdate(newLeaf); err != nil {
+		return nil, err
+	}
+	newParentNode := NewParentNode()
+	newParentNode.SetChild(pathNewLeaf.Get(lvl), newLeaf)
+	newParentNode.SetChild(pathNewLeaf.GetOther(lvl), oldLeaf)
+	return newParentNode, t.checkNodeHashConflictDuringUpdate(newParentNode)
+}
+
+func (t *MerkleTree) TryDelete(key []byte) error {
+	k, err := zkt.ToSecureKey(key)
+	if err != nil {
+		return err
+	}
+	return t.TryDeleteByNodeKey(zkt.NewHashFromBigInt(k))
+}
+
+// TryDeleteByNodeKey removes the specified Key from the MerkleTree and updates the path
+// from the deleted key to the Root with the new values. This method removes
+// the key from the MerkleTree, but does not remove the old nodes from the
+// key-value database; this means that if the tree is accessed by an old Root
+// where the key was not deleted yet, the key will still exist. If is desired
+// to remove the key-values from the database that are not under the current
+// Root, an option could be to dump all the leafs (using mt.DumpLeafs) and
+// import them in a new MerkleTree in a new database (using
+// mt.ImportDumpedLeafs), but this will lose all the Root history of the MerkleTree
+func (t *MerkleTree) TryDeleteByNodeKey(nodeKey *zkt.Hash) error {
+	if !zkt.CheckBigIntInField(nodeKey.BigInt()) { // verify that k is valid and fit inside the Finite Field.
+		return trie.ErrInvalidField
+	}
+	node, path, pathNodes := t.rootNode, t.newTreePath(nodeKey), *new([]*ParentNode)
+	for i := 0; i < t.maxLevels; i++ {
+		switch n := node.(type) {
+		case *ParentNode:
+			pathNodes = append(pathNodes, n)
+			node = t.getChild(n, path.Get(i))
+			continue
+		case *LeafNode:
+			if bytes.Equal(nodeKey[:], n.KeyHash[:]) {
+				t.rmAndUpload(path, pathNodes)
+				return nil
+			}
+			return trie.ErrKeyNotFound
+		case *EmptyNode:
+			return trie.ErrKeyNotFound
+		case *HashNode:
+			return trie.ErrKeyNotFound
+		default:
+			return trie.ErrInvalidNodeFound
+		}
+	}
+	return trie.ErrKeyNotFound
+}
+
+// rmAndUpload removes the key, and goes up until the root updating all the nodes with the new values.
+func (t *MerkleTree) rmAndUpload(path TreePath, pathNodes []*ParentNode) {
+	switch len(pathNodes) {
+	case 0: // root node does not parent node.
+		t.removeNodeHash(t.rootNode)
+		t.rootNode = EmptyNodeValue
+	case 1:
+		t.removeNodeHash(t.rootNode)
+		t.rootNode = t.getChild(pathNodes[0], path.GetOther(0))
```

Comment:
```suggestion
	case left: // root node does not parent node.
		t.removeNodeHash(t.rootNode)
		t.rootNode = EmptyNodeValue
	case right:
		t.removeNodeHash(t.rootNode)
		t.rootNode = t.getChild(pathNodes[0], path.GetOther(0))
```

Reply:
That's not a tree path, that's a tree depth.

Reply:
Aha sorry about that!